### PR TITLE
NumericUpDown control changes

### LINKED ISSUE #5318: NumericUpDown control
**Describe the bug**
When the NumericUpDown control is at a value of 5.6 and Increment is 0.01, if adjusted up, it will appear as 0.570000000000000001 instead of 0.57

**Code**
`<NumericUpDown Value="0.56" Minimum="0" Maximum="10" Increment="0.01"/>`

**Expected behavior**
The value should beThe value should be 0.57

**Screenshots**
image
image

**Desktop (please complete the following information):**
 - OS: win 10
 - Version  0.9.12

### PR DESCRIPTION
**Pull request by puppetsw**
## What does the pull request do?

Attempts to issue https://github.com/AvaloniaUI/Avalonia/issues/5318 fix by adding a property for decimal places to the control.

## What is the current behavior?

When the NumericUpDown control is at a value of 5.6 and Increment is 0.01, if adjusted up, it will appear as 0.570000000000000001 instead of 0.57

## What is the updated/expected behavior with this PR?

The value should be 0.57

## How was the solution implemented (if it's not obvious)?
Changing the backing type for NumericUpDown to decimal breaks contracts.

## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

Changing the backing type for NumericUpDown to decimal breaks contracts.

## Obsoletions / Deprecations
CultureInfo property is obsolete @AvaloniaUI/core 
Users should use NumberFormat instead.

## Fixed issues
Fixes #5318

### CODE REVIEW

**Comment by maxkatz6** [comment]
Thanks for the contribution!
Am I right, that with this implementation if won't apply rounding for values that were manually inputed or came from binding? So, everything but increment/decrement.
I had quick look, and it seems that ConvertTextToValue method suits well to have rounding: https://github.com/AvaloniaUI/Avalonia/blob/2956f79c5c8268dd4d3f763324a501336e756394/src/Avalonia.Controls/NumericUpDown/NumericUpDown.cs#L607

**Comment by danwalmsley** [comment]
@maxkatz6 probably we have something in MathUtils class that can deal with this kind of stuff?

**Comment by maxkatz6** [comment]
@danwalmsley I don't see anything that can help us in https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Utilities/MathUtilities.cs

But I am not sure, do we need to Round here or Truncate?

**Comment by robloo** [comment]
Internally, the backing type should probably be `decimal`. Then `Value` is just a double conversion of the internal decimal for backwards compatiblity. A new `DecimalValue` should be added. The only reason this wasn't done in the NumberBox in WinUI is because C++ has no `decimal` type.

**Comment by puppetsw** [comment]
> Thanks for the contribution!
> Am I right, that with this implementation if won't apply rounding for values that were manually inputed or came from binding? So, everything but increment/decrement.
> I had quick look, and it seems that ConvertTextToValue method suits well to have rounding:
> 
> https://github.com/AvaloniaUI/Avalonia/blob/2956f79c5c8268dd4d3f763324a501336e756394/src/Avalonia.Controls/NumericUpDown/NumericUpDown.cs#L607
Yes, that was how I approached it. But it makes more sense to change the backing type to decimal considering currency input. 

I'm happy to make the changes for decimal backing type. or Adding a new DecimalValue. 

Sorry I'm new to this, should I close out this PR?

**Comment by robloo** [comment]
A few other comments:
 - If it was me, I wouldn't close the PR. I would just revert commits and add changes as needed. It's good to keep all the discussion like this in one place.
 - Generally, rounding should be handled by the number formatting itself. I would never add a `DecimalPlaces` property like some old controls might have. Instead, the ToString and parsing methods should be using a `NumberFormatInfo` that can be defined by the developer as a property. Otherwise, it should default to the current UI culture one. This is an oversimplification of a much larger discussion that was had over on the WinUI repo (still unresolved). Bottom line, Hex, Currency, Decimal, Integer, etc formats all need to be easily supported. This at first seems out of scope for this discussion, but is very much related to decimal place precision as it relates to overall formatting.

**Comment by maxkatz6** [comment]
@puppetsw technically switching from double to decimal is a breaking changes, that's why build is failing. You need to follow this tutorial to allow that https://stu.dev/check-for-breaking-changes-with-apicompat/ .

**Comment by puppetsw** [comment]
I've made type changes to the NumericUpDown class and event class. Added a NumberFormat property and have also set it through the CultureInfo to keep backwards compatibility. Marked CultureInfo as Obsolete. 

I also added a new Clamp method to the MathUtilities.

**Comment by robloo** [comment]
> Marked CultureInfo as Obsolete.

Core team doesn't really like anyone marking things as obsolete unless you run in past them first. @danwalmsley will need to buy in on this.

File: src/Avalonia.Base/Utilities/MathUtilities.cs [hunk 1]
```diff
@@ -224,6 +224,34 @@ public static double Clamp(double val, double min, double max)
...
         }
 
+        /// <summary>
+        /// Clamps a value between a minimum and maximum value.
+        /// </summary>
+        /// <param name="val">The value.</param>
+        /// <param name="min">The minimum value.</param>
+        /// <param name="max">The maximum value.</param>
+        /// <returns>The clamped value.</returns>
+        public static decimal Clamp(decimal val, decimal min, decimal max)
```
Comment (pr8x) [inline src/Avalonia.Base/Utilities/MathUtilities.cs:234]: What's wrong with `System.Math.Clamp(Decimal, Decimal, Decimal)` ? https://docs.microsoft.com/en-us/dotnet/api/system.math.clamp?view=net-5.0#System_Math_Clamp_System_Decimal_System_Decimal_System_Decimal_

Comment (maxkatz6) [reply src/Avalonia.Base/Utilities/MathUtilities.cs:234]: It's not available in netstandard 2.0

**Comment by grokys** [comment]
I've not really been following this PR - are we agreed that it's a good idea to break the API and change the values from `double` to `decimal`?

**Comment by maxkatz6** [comment]
@grokys personally I agree with that. Also it seems to be binary breaking change only, since both xaml and c# implicitly convert from decimal to double and in other way.

**Comment by grokys** [comment]
Ok, I'm fine with accepting your judgement on that ;) Looks good to me, but I'll leave the final review to you.

**Review by maxkatz6 — APPROVED** [review]

**Comment by maxkatz6** [comment]
Updated samples and ApiCompat. Thanks!

**Pull request merged**